File: laughodds_mobile/lib/services/yellow_network_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:web_socket_channel/web_socket_channel.dart';
import '../config/constants.dart';

/// Yellow Network (Nitrolite) state channel service.
/// Connects to ClearNode via WebSocket for auth, balance, and vote transfers.
class YellowNetworkService {
  WebSocketChannel? _channel;
  bool _isConnected = false;
  bool _isAuthenticated = false;
  String _balance = '0';
  String? _channelId;
  String? _error;
  int _reqId = 1;

  final _pendingTransfers = <int, Completer<Map<String, dynamic>>>{};
  final _statusController = StreamController<void>.broadcast();

  // Public getters
  bool get isConnected => _isConnected;
  bool get isAuthenticated => _isAuthenticated;
  String get balance => _balance;
  String? get channelId => _channelId;
  String? get error => _error;
  Stream<void> get statusStream => _statusController.stream;

  int _nextReqId() => _reqId++;

  /// Build Nitrolite RPC request
  Map<String, dynamic> _buildRequest(String method, Map<String, dynamic> params) {
    return {
      'req': [_nextReqId(), method, params, DateTime.now().millisecondsSinceEpoch],
      'sig': <String>[],
    };
  }

  /// Parse Nitrolite RPC response
  Map<String, dynamic>? _parseResponse(String data) {
    try {
      final msg = jsonDecode(data);
      if (msg['res'] != null && msg['res'] is List && (msg['res'] as List).length >= 3) {
        final res = msg['res'] as List;
        return {'id': res[0], 'method': res[1], 'result': res[2]};
      }
      return null;
    } catch (_) {
      return null;
    }
  }

  /// Connect to ClearNode WebSocket
  Future<void> connect(String walletAddress) async {
    if (_isConnected) return;

    try {
      debugPrint('[Yellow] Connecting to ClearNode...');
      _channel = WebSocketChannel.connect(Uri.parse(clearnodeWsUrl));

      await _channel!.ready;
      _isConnected = true;
      _error = null;
      _notifyStatus();
      debugPrint('[Yellow] WebSocket connected');

      // Listen for messages
      _channel!.stream.listen(
        (data) => _handleMessage(data.toString(), walletAddress),
        onError: (err) {
          debugPrint('[Yellow] WebSocket error: $err');
          _error = 'Yellow Network connection error';
          _isConnected = false;
          _isAuthenticated = false;
          _notifyStatus();
        },
        onDone: () {
          debugPrint('[Yellow] WebSocket disconnected');
          _isConnected = false;
          _isAuthenticated = false;
          _channel = null;
          _notifyStatus();
        },
      );

      // Start auth
      _startAuth(walletAddress);
    } catch (e) {
      debugPrint('[Yellow] Connection failed: $e');
      _error = 'Failed to connect to Yellow Network';
      _isConnected = false;
      _notifyStatus();
    }
  }

  /// Start authentication flow
  void _startAuth(String walletAddress) {
    // For mobile (no MetaMask), we send auth_request
    // The server-side handles the session key approach
    // Mobile uses a simplified auth: just declare identity
    final params = {
      'address': walletAddress,
      'session_key': walletAddress, // Use wallet address as session key for mobile
      'application': 'ViralForge',
      'allowances': [
        {'asset': voteAsset, 'amount': '1000000000'}
      ],
      'expires_at': (DateTime.now().millisecondsSinceEpoch ~/ 1000) + 3600,
      'scope': 'test.app',
    };

    _sendMessage(_buildRequest('auth_request', params));
  }

  /// Handle incoming WebSocket messages
  void _handleMessage(String data, String walletAddress) {
    final parsed = _parseResponse(data);
    if (parsed == null) {
      debugPrint('[Yellow] Unparseable message: $data');
      return;
    }

    final id = parsed['id'] as int;
    final method = parsed['method'] as String;
    final result = parsed['result'];

    debugPrint('[Yellow] Received: $method');

    switch (method) {
      case 'auth_challenge':
        // For mobile without EIP-712 signing, we attempt a simplified verify
        // In production, this would need proper wallet signing
        final challenge = result?['challenge_message'] as String?;
        if (challenge != null) {
          _sendMessage({
            'req': [
              _nextReqId(),
              'auth_verify',
              {'challenge': challenge},
              DateTime.now().millisecondsSinceEpoch,
            ],
            'sig': ['0x'], // Placeholder - mobile simplified auth
          });
        }
        break;

      case 'auth_verify':
        if (result?['success'] == true ||
            result?['authenticated'] == true ||
            result?['session_key'] != null) {
          debugPrint('[Yellow] Authenticated!');
          _isAuthenticated = true;
          _error = null;
          // Fetch balance and channels
          _sendMessage(_buildRequest('get_ledger_balances', {}));
          _sendMessage(_buildRequest('get_channels', {}));
          // Request faucet
          requestFaucetTokens(walletAddress);
        } else {
          debugPrint('[Yellow] Auth failed: $result');
          _error = result?['error']?.toString() ?? 'Authentication failed';
          _isAuthenticated = false;
        }
        _notifyStatus();
        break;

      case 'get_ledger_balances':
        final entries = result?['ledger_balances'] as List? ?? [];
        for (final entry in entries) {
          if (entry['asset'] == voteAsset) {
            final amount = double.tryParse(entry['amount']?.toString() ?? '0') ?? 0;
            _balance = (amount / 1000000).toString();
          }
        }
        _notifyStatus();
        break;

      case 'bu': // Balance update push notification
        final updates = result?['balance_updates'] as List? ?? [];
        for (final entry in updates) {
          if (entry['asset'] == voteAsset) {
            final amount = double.tryParse(entry['amount']?.toString() ?? '0') ?? 0;
            _balance = (amount / 1000000).toString();
          }
        }
        if (updates.isEmpty) {
          _sendMessage(_buildRequest('get_ledger_balances', {}));
        }
        _notifyStatus();
        break;

      case 'channels':
      case 'get_channels':
        final channelList = result?['channels'] as List? ??
            (result is List ? result : []);
        for (final ch in channelList) {
          if (ch['status'] == 'open') {
            _channelId = ch['channel_id']?.toString() ?? ch['id']?.toString();
            break;
          }
        }
        _notifyStatus();
        break;

      case 'cu': // Channel update notification
        _sendMessage(_buildRequest('get_channels', {}));
        break;

      case 'tr': // Transfer notification
      case 'asu': // App session update
        break;

      case 'open_channel':
        if (result?['channel_id'] != null) {
          _channelId = result['channel_id'].toString();
          debugPrint('[Yellow] Channel opened: $_channelId');
        }
        _notifyStatus();
        break;

      case 'transfer':
        final completer = _pendingTransfers.remove(id);
        if (completer != null) {
          final txs = result?['transactions'] as List?;
          if (txs != null && txs.isNotEmpty) {
            completer.complete({
              'success': true,
              'transferId': txs[0]['id']?.toString() ?? '',
            });
            _sendMessage(_buildRequest('get_ledger_balances', {}));
          } else if (result?['error'] != null) {
            completer.complete({
              'success': false,
              'error': result['error'].toString(),
            });
          } else {
            completer.complete({
              'success': false,
              'error': 'Transfer failed - no transaction returned',
            });
          }
        }
        break;

      case 'error':
        debugPrint('[Yellow] Server error: ${result?['error']}');
        _error = result?['error']?.toString();
        _notifyStatus();
        break;

      default:
        debugPrint('[Yellow] Unhandled: $method');
    }
  }

  /// Send a vote transfer via state channel
  Future<Map<String, dynamic>> sendVoteTransfer({
    required String toAddress,
    required int marketId,
    required String vote,
  }) async {
    if (!_isAuthenticated) {
      return {'success': false, 'error': 'Not authenticated with Yellow Network'};
    }

    final reqId = _nextReqId();
    final params = {
      'allocations': [
        {'asset': voteAsset, 'amount': voteCostAmount}
      ],
      'destination': toAddress,
    };

    final req = [reqId, 'transfer', params, DateTime.now().millisecondsSinceEpoch];
    final completer = Completer<Map<String, dynamic>>();
    _pendingTransfers[reqId] = completer;

    _channel?.sink.add(jsonEncode({'req': req, 'sig': <String>[]}));

    // Timeout after 30s
    return completer.future.timeout(
      const Duration(seconds: 30),
      onTimeout: () {
        _pendingTransfers.remove(reqId);
        return {'success': false, 'error': 'Transfer timeout'};
      },
    );
  }

  /// Refresh balance
  void refreshBalance() {
    if (_isAuthenticated) {
      _sendMessage(_buildRequest('get_ledger_balances', {}));
    }
  }

  /// Request faucet tokens from Yellow Network
  Future<bool> requestFaucetTokens(String walletAddress) async {
    try {
      final response = await http.post(
        Uri.parse(yellowFaucetUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'userAddress': walletAddress}),
      );
      return response.statusCode == 200;
    } catch (e) {
      debugPrint('[Yellow] Faucet request failed: $e');
      return false;
    }
  }

  /// Disconnect from ClearNode
  void disconnect() {
    _channel?.sink.close();
    _channel = null;
    _isConnected = false;
    _isAuthenticated = false;
    _channelId = null;
    _balance = '0';
    _notifyStatus();
  }

  void _sendMessage(Map<String, dynamic> msg) {
    if (_channel != null && _isConnected) {
      _channel!.sink.add(jsonEncode(msg));
    }
  }

  void _notifyStatus() {
    if (!_statusController.isClosed) {
      _statusController.add(null);
    }
  }

  void dispose() {
    disconnect();
    _statusController.close();
  }
}

File: server/services/yellowNetworkService.js
// server/services/yellowNetworkService.js
// Server-side Yellow Network state channel management
const WebSocket = require('ws');

const CLEARNODE_WS_URL = process.env.CLEARNODE_WS_URL || 'wss://clearnet-sandbox.yellow.com/ws';
const YELLOW_PRIVATE_KEY = process.env.YELLOW_PRIVATE_KEY;
const VOTE_COST_AMOUNT = '100'; // 0.0001 ytest.usd in base units

class YellowNetworkService {
    constructor() {
        this.ws = null;
        this.isConnected = false;
        this.isAuthenticated = false;
        this.sessionKey = null;
        this.balance = '0';
        this.nonce = 0;
        this.pendingTransfers = new Map();
        this.reconnectTimer = null;
    }

    // Generate server session key from private key
    generateSessionKey() {
        if (!YELLOW_PRIVATE_KEY) {
            throw new Error('YELLOW_PRIVATE_KEY not configured');
        }
        // Use a deterministic session key derived from the private key
        const crypto = require('crypto');
        const hash = crypto.createHash('sha256').update(YELLOW_PRIVATE_KEY + Date.now()).digest('hex');
        return '0x' + hash;
    }

    // Connect to Clearnode
    async connect() {
        return new Promise((resolve, reject) => {
            if (this.ws && this.isConnected) {
                resolve();
                return;
            }

            console.log('[Yellow Server] Connecting to Clearnode:', CLEARNODE_WS_URL);

            try {
                this.ws = new WebSocket(CLEARNODE_WS_URL);
            } catch (err) {
                console.error('[Yellow Server] Failed to create WebSocket:', err.message);
                reject(err);
                return;
            }

            const timeout = setTimeout(() => {
                reject(new Error('Connection timeout'));
            }, 30000);

            this.ws.on('open', () => {
                clearTimeout(timeout);
                console.log('[Yellow Server] WebSocket connected');
                this.isConnected = true;
                this.startAuth(resolve, reject);
            });

            this.ws.on('message', (data) => {
                this.handleMessage(data, resolve, reject);
            });

            this.ws.on('error', (err) => {
                console.error('[Yellow Server] WebSocket error:', err.message);
                this.isConnected = false;
                this.isAuthenticated = false;
            });

            this.ws.on('close', () => {
                console.log('[Yellow Server] WebSocket disconnected');
                this.isConnected = false;
                this.isAuthenticated = false;
                this.ws = null;

                // Reconnect after 5 seconds
                if (!this.reconnectTimer) {
                    this.reconnectTimer = setTimeout(() => {
                        this.reconnectTimer = null;
                        this.connect().catch(err => {
                            console.error('[Yellow Server] Reconnect failed:', err.message);
                        });
                    }, 5000);
                }
            });
        });
    }

    startAuth(resolve, reject) {
        if (!YELLOW_PRIVATE_KEY) {
            console.warn('[Yellow Server] No YELLOW_PRIVATE_KEY set, skipping auth');
            resolve();
            return;
        }

        this.sessionKey = this.generateSessionKey();
        this._authResolve = resolve;
        this._authReject = reject;

        this.send({
            type: 'auth_request',
            session_key: this.sessionKey,
            wallet_address: this.getServerAddress(),
            timestamp: Date.now(),
        });
    }

    getServerAddress() {
        if (!YELLOW_PRIVATE_KEY) return '0x0000000000000000000000000000000000000000';
        // Derive address from private key using ethers
        try {
            const { ethers } = require('ethers');
            const wallet = new ethers.Wallet(YELLOW_PRIVATE_KEY);
            return wallet.address;
        } catch (err) {
            console.error('[Yellow Server] Failed to derive address:', err.message);
            return '0x0000000000000000000000000000000000000000';
        }
    }

    async handleMessage(data, authResolve, authReject) {
        try {
            const msg = JSON.parse(data.toString());
            console.log('[Yellow Server] Received:', msg.type);

            switch (msg.type) {
                case 'auth_challenge': {
                    // Sign with server private key
                    try {
                        const { ethers } = require('ethers');
                        const wallet = new ethers.Wallet(YELLOW_PRIVATE_KEY);

                        // Sign the challenge
                        const signature = await wallet.signTypedData(
                            {
                                name: msg.domain?.name || 'Yellow Network',
                                version: msg.domain?.version || '1',
                                chainId: msg.domain?.chainId || 11155111,
                            },
                            {
                                AuthMessage: [
                                    { name: 'challenge', type: 'string' },
                                    { name: 'session_key', type: 'address' },
                                ],
                            },
                            {
                                challenge: msg.challenge,
                                session_key: this.sessionKey,
                            }
                        );

                        this.send({
                            type: 'auth_verify',
                            wallet_address: wallet.address,
                            session_key: this.sessionKey,
                            signature,
                            challenge: msg.challenge,
                        });
                    } catch (signErr) {
                        console.error('[Yellow Server] Auth signing failed:', signErr.message);
                        if (authReject) authReject(signErr);
                    }
                    break;
                }

                case 'auth_success': {
                    console.log('[Yellow Server] Authenticated successfully');
                    this.isAuthenticated = true;
                    this.send({ type: 'get_balance' });
                    if (authResolve) authResolve();
                    break;
                }

                case 'auth_error': {
                    console.error('[Yellow Server] Auth error:', msg.message);
                    this.isAuthenticated = false;
                    if (authReject) authReject(new Error(msg.message));
                    break;
                }

                case 'balance': {
                    this.balance = msg.balances?.['ytest.usd'] || '0';
                    console.log('[Yellow Server] Balance:', this.balance, 'ytest.usd');
                    break;
                }

                case 'transfer_success': {
                    const pending = this.pendingTransfers.get(msg.transfer_id);
                    if (pending) {
                        pending.resolve({ success: true, transferId: msg.transfer_id });
                        this.pendingTransfers.delete(msg.transfer_id);
                    }
                    this.send({ type: 'get_balance' });
                    break;
                }

                case 'transfer_error': {
                    const pending = this.pendingTransfers.get(msg.transfer_id);
                    if (pending) {
                        pending.reject(new Error(msg.message || 'Transfer failed'));
                        this.pendingTransfers.delete(msg.transfer_id);
                    }
                    break;
                }

                case 'transfer_received': {
                    // Incoming transfer from user (vote payment)
                    console.log('[Yellow Server] Received transfer:', msg);
                    break;
                }

                default:
                    console.log('[Yellow Server] Unhandled:', msg.type);
            }
        } catch (err) {
            console.error('[Yellow Server] Error handling message:', err.message);
        }
    }

    send(msg) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(msg));
        } else {
            console.warn('[Yellow Server] Cannot send, not connected');
        }
    }

    // Validate incoming vote transfer from user
    validateIncomingVoteTransfer(transferProof) {
        if (!transferProof || !transferProof.transferId) {
            return { valid: false, error: 'Missing transfer proof' };
        }

        // In production, verify the transfer proof cryptographically
        // For now, trust the transferId and validate via Clearnode API
        return { valid: true };
    }

    // Distribute rewards to winners via Yellow Network
    async distributeRewards(winners) {
        if (!this.isAuthenticated) {
            throw new Error('Not authenticated with Yellow Network');
        }

        const results = [];

        for (const winner of winners) {
            try {
                const transferId = `reward_${winner.address}_${Date.now()}`;
                const result = await this.sendTransfer(
                    winner.address,
                    winner.amount.toString(),
                    transferId
                );
                results.push({ address: winner.address, success: true, transferId });
                console.log(`[Yellow Server] Reward sent to ${winner.address}: ${winner.amount} ytest.usd`);
            } catch (err) {
                console.error(`[Yellow Server] Failed to send reward to ${winner.address}:`, err.message);
                results.push({ address: winner.address, success: false, error: err.message });
            }
        }

        return results;
    }

    // Send a transfer via Yellow Network
    sendTransfer(toAddress, amount, transferId) {
        return new Promise((resolve, reject) => {
            if (!this.isAuthenticated) {
                reject(new Error('Not authenticated'));
                return;
            }

            this.pendingTransfers.set(transferId, { resolve, reject });

            this.send({
                type: 'transfer',
                transfer_id: transferId,
                to: toAddress,
                asset: 'ytest.usd',
                amount,
                nonce: ++this.nonce,
            });

            // Timeout after 30s
            setTimeout(() => {
                if (this.pendingTransfers.has(transferId)) {
                    this.pendingTransfers.delete(transferId);
                    reject(new Error('Transfer timeout'));
                }
            }, 30000);
        });
    }

    getBalance() {
        return this.balance;
    }

    getStatus() {
        return {
            isConnected: this.isConnected,
            isAuthenticated: this.isAuthenticated,
            balance: this.balance,
        };
    }
}

module.exports = new YellowNetworkService();
